src/rt-app_parse_config.c

```diff
@@ -822,6 +823,64 @@ static sched_data_t *parse_sched_data(struct json_object *obj, int def_policy)
 	return NULL;
 }
 
+static taskgroup_data_t *parse_taskgroup_data(struct json_object *obj)
+{
+	char *name = get_string_value_from(obj, "taskgroup", TRUE, ""), *curr;
+	unsigned int len = strlen(name);
+	taskgroup_data_t *tg;
+
+	if (!len)
+		return NULL;
+
+	tg = find_taskgroup(name);
+	if (tg)
+		return tg;
+
+	tg = alloc_taskgroup(len + 1);
```

Comment:
You allocate the memory for the name in alloc_taskgroup but then you check some char must be removed. This looks like a waste of memory.
compute what will be the final size of the name
allocate memory
set the name

Reply:
That's true. Since 'computing the size' and 'setting the name' are very similar operations, I factored out 'size_t parse_taskgroup_name(char *src, char *dst, bool copy)' which I use for both operations (with copy = 0 and copy = 1 accordingly).

``` diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index 568f20f351ba..18bc62cb9aeb 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -823,38 +823,57 @@ static sched_data_t *parse_sched_data(struct json_object *obj, int def_policy)
        return NULL;
 }
 
+static size_t parse_taskgroup_name(char *src, char *dst, int copy)
+{
+       size_t len = 1;
+
+       /* Enforce leading slash. */
+       *dst = '/';
+
+       for (; *src != '\0'; src++)
+               /* Reduce sequence of slashes to one slash. */
+               if (*src != '/' || *src != *dst) {
+                       if (copy)
+                               dst++;
+                       *dst = *src;
+                       len++;
+               }
+
+       /* Remove trailing slash except for root taskgroup. */
+       if (len == 1 || *dst != '/') {
+               if (copy)
+                       dst++;
+               len++;
+       }
+
+       *dst = '\0';
+
+       return len;
+}
+
 static taskgroup_data_t *parse_taskgroup_data(struct json_object *obj)
 {
-       char *name = get_string_value_from(obj, "taskgroup", TRUE, ""), *curr;
-       unsigned int len = strlen(name);
+       char *name = get_string_value_from(obj, "taskgroup", TRUE, "");
        taskgroup_data_t *tg;
+       size_t size;
+       char dummy;
 
-       if (!len)
+       if (!strlen(name))
                return NULL;
 
        tg = find_taskgroup(name);
        if (tg)
                return tg;
 
-       tg = alloc_taskgroup(len + 1);
+       size = parse_taskgroup_name(name, &dummy, 0);
+
+       tg = alloc_taskgroup(size);
        if (!tg) {
                log_critical(PIN2 "Cannot allocate taskgroup");
                exit(EXIT_FAILURE);
        }
 
-       curr = tg->name;
-       /* Enforce leading slash. */
-       *curr = '/';
-
-       for (; *name != '\0'; name++)
-               /* Reduce sequence of slashes to one slash. */
-               if (*name != '/' || *name != *curr)
-                       *(++curr) = *name;
-
-       /* Remove trailing slash except for root taskgroup. */
-       if (curr == tg->name || *curr != '/')
-               curr++;
-       *curr = '\0';
+       parse_taskgroup_name(name, tg->name, 1);
 
        return tg;
 }```

Reply:
Apart from adding a comment explaining why you keep updating dst with *dst = *src when computing size (copy == false), this looks good to me

Reply:
OK, I can add:

```
diff --git a/src/rt-app_parse_config.c b/src/rt-app_parse_config.c
index a19297d42266..d04aed0f3609 100644
--- a/src/rt-app_parse_config.c
+++ b/src/rt-app_parse_config.c
@@ -835,6 +835,10 @@ static size_t parse_taskgroup_name(char *src, char *dst, int copy)
                if (*src != '/' || *src != *dst) {
                        if (copy)
                                dst++;
+                       /*
+                        * Set *dst to *src even in 'copy == 0' case for
+                        * 'sequence of of slashes' detection.
+                        */
                        *dst = *src;
                        len++;
                }
```

Reply:
looks good to me